lib/strings: call toString within hasInfix

danth opened the pull request:
###### Description of changes

This change should fix [this issue](https://github.com/NixOS/nixpkgs/pull/168175#issuecomment-1107830574), and any similar problems which may arise.
It also allows `hasInfix` to be used over derivations as was previously possible.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

pennae commented on lib/strings.nix:
```diff
@@ -253,7 +253,7 @@ rec {
   */
   hasInfix = infix: content:
-    builtins.match ".*${escapeRegex infix}.*" content != null;
+    builtins.match ".*${escapeRegex infix}.*" (toString content) != null;
```
that's probably a bit broad since toString will take just about anything. how about
```suggestion
    builtins.match ".*${escapeRegex infix}.*" (if builtins.isPath content or isDerivation content then toString content else content) != null;
```

pennae replied:
... which is probably not broad enough because `__toString` exists. :facepalm: 
```suggestion
    let content' =
      if builtins.isPath content || isDerivation content || content ? __toString
      then toString content
      else content;
    in
      builtins.match ".*${escapeRegex infix}.*" content' != null;
```

danth committed:
lib/strings: convert to string within hasInfix

This should fix the issue mentioned here:
https://github.com/NixOS/nixpkgs/pull/168175#issuecomment-1107830574

danth commented:
@pennae Implemented both of your suggestions.

I think we have everything that was accepted by the old function now.

pennae commented:
yeah, would think so too. @Izorkin can you test with nixops?

Izorkin commented:
Yes, working!

roberth reviewed (changes requested):
This is still different from the original. It's going to need a bunch of test cases. You can add those in `lib/tests/misc.nix`.

tomberek commented:
Call toString unconditionally?

pennae commented:
that would admit way too much, eg `hasInfix "1" 2010`. apart from plugin-provided value types this new definition covers all the bases, though we could use `substring 0 9223372036854775808 content` instead of `toString` to get perfectly equivalent coercion behaviour?

roberth commented:
> Call toString unconditionally?

Nope.
Test it and you'll find the answer, and have a tests. A nice challenge, with a nice outcome, I figured.
I'll be happy to help if your tests do not find it.

roberth commented on lib/tests/misc.nix:
```diff
@@ -251,6 +251,31 @@ runTests {
+    expr = hasInfix "tests" ./.;
+    expected = true;
+  };
+
```
```suggestion
  
  testHasInfixPathStoreDir = {
    expr = hasInfix builtins.storeDir ./.;
    expected = true;
  };

```

String coercions in Nix are tricky.

danth replied:
This test is failing. How do we fix it?

danth committed:
lib/tests: add tests for hasInfix

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

tomberek commented:
Trying to recover full backward compat might involve enough cases that the original perf rationale goes away. Should we revert until all this is figured out?

roberth commented:
Maybe `"${x}"` is the right string coercion?
Nix has more variations on string coercion than I'd like, but it's finite.

danth commented:
@roberth Seems like it.

Something went wrong with my rebase causing some extra commits to get added, will fix that now.

roberth commented:
Thanks @danth!
